src/components/Output/index.tsx

```diff
@@ -44,62 +44,88 @@ export default class Output extends Component<Props, State> {
   retargetedEvents = new WeakSet<Event>();
 
   componentDidMount() {
-    if (this.canvasLeft) {
-      drawDataToCanvas(this.canvasLeft, this.props.leftImg);
+    const leftDraw = this.leftDrawable();
+    const rightDraw = this.rightDrawable();
+
+    if (this.canvasLeft && leftDraw) {
+      drawDataToCanvas(this.canvasLeft, leftDraw);
     }
-    if (this.canvasRight) {
-      drawDataToCanvas(this.canvasRight, this.props.rightImg);
+    if (this.canvasRight && rightDraw) {
+      drawDataToCanvas(this.canvasRight, rightDraw);
     }
   }
 
   componentDidUpdate(prevProps: Props, prevState: State) {
-    if (prevProps.leftImg !== this.props.leftImg && this.canvasLeft) {
-      drawDataToCanvas(this.canvasLeft, this.props.leftImg);
+    const prevLeftDraw = this.leftDrawable(prevProps);
+    const prevRightDraw = this.rightDrawable(prevProps);
+    const leftDraw = this.leftDrawable();
+    const rightDraw = this.rightDrawable();
+
+    if (leftDraw && leftDraw !== prevLeftDraw && this.canvasLeft) {
+      drawDataToCanvas(this.canvasLeft, leftDraw);
     }
-    if (prevProps.rightImg !== this.props.rightImg && this.canvasRight) {
-      drawDataToCanvas(this.canvasRight, this.props.rightImg);
+    if (rightDraw && rightDraw !== prevRightDraw && this.canvasRight) {
+      drawDataToCanvas(this.canvasRight, rightDraw);
+    }
+
+    if (this.props.originalImage !== prevProps.originalImage && this.pinchZoomLeft) {
+      // New image? Reset the pinch-zoom.
+      this.pinchZoomLeft.setTransform({
+        allowChangeEvent: true,
+        x: 0,
+        y: 0,
+        scale: 1,
+      });
     }
   }
 
   shouldComponentUpdate(nextProps: Props, nextState: State) {
     return !shallowEqual(this.props, nextProps) || !shallowEqual(this.state, nextState);
   }
 
+  private leftDrawable(props: Props = this.props): ImageData | undefined {
+    return props.leftCompressed || props.originalImage;
+  }
+
+  private rightDrawable(props: Props = this.props): ImageData | undefined {
+    return props.rightCompressed || props.originalImage;
+  }
+
   @bind
-  toggleBackground() {
+  private toggleBackground() {
```

Comment:
I'm making all internal methods private, as it makes it easier to spot methods that aren't used.